src/signature.rs

```diff
@@ -139,12 +139,15 @@ pub fn verify(signature: &Signature, hashes: &[G2Projective], public_keys: &[Pub
     // Enforce that messages are distinct as a countermeasure against BLS's rogue-key attack.
     // See Section 3.1. of the IRTF's BLS signatures spec:
     // https://tools.ietf.org/html/draft-irtf-cfrg-bls-signature-02#section-3.1
-    for i in 0..(n_hashes - 1) {
-        for j in (i + 1)..n_hashes {
-            if hashes[i] == hashes[j] {
-                return false;
-            }
-        }
+    if hashes
+        .iter()
+        // This is the best way to get something we can actually hash.
+        .map(|h| G2Affine::from(h).to_compressed())
```

Comment:
Maybe using `G2Projective::to_bytes` (which does exactly the same thing as what you added) would make it totally clear that you're comparing bytes, and that the point representation (affine/projective compressed/uncompressed) doesn't affect duplicate checking?
```rust
// Convert G2 point to a type that implements `Hash`.
.map(G2Projective::to_bytes)
```

Reply:
Unfortunately, that returns a `G2Compressed` which doesn't let me get at the underlying `[u8; 96]`. I can call `as_ref` on that, but then I'd need to copy the bytes out into a new buffer and deal with all the potential error cases (and/or allocate).

I'm happy to do that if you think it'll work better, but it'll end up being:

```rust
<[u8; 96]>::try_from(h.to_bytes().as_ref()).unwrap()
```

Reply:
I'll see if I can get upstream to accept an `Into<[u8; 96]>` patch.

Reply:
https://github.com/zkcrypto/bls12_381/pull/119. But in the meantime, I'm happy to do the copy dance if you think it's easier to understand.

Reply:
Ha oh right, `to_bytes` doesn't actually return the array of bytes... so I think how it's currently written is best.